File: clients/haskell/src/Model/UnitOrder.hs
module Model.UnitOrder where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Model.ActionOrder (ActionOrder)
import Model.Vec2 (Vec2)

-- | Order for specific unit
data UnitOrder = UnitOrder {
    -- | Target moving velocity
    targetVelocity :: Vec2,
    -- | Target view direction (vector length doesn't matter)
    targetDirection :: Vec2,
    -- | Order to perform an action, or None
    action :: Maybe ActionOrder }
    deriving Show

instance Trans UnitOrder where
    read = do
        targetVelocity <- Trans.read
        targetDirection <- Trans.read
        action <- Trans.read
        return UnitOrder {
            targetVelocity,
            targetDirection,
            action }
    
    write UnitOrder {
        targetVelocity,
        targetDirection,
        action } = do
            Trans.write targetVelocity
            Trans.write targetDirection
            Trans.write action

File: clients/rust/src/model/unit_order.rs
use super::*;

/// Order for specific unit
#[derive(Clone, Debug)]
pub struct UnitOrder {
    /// Target moving velocity
    pub target_velocity: model::Vec2,
    /// Target view direction (vector length doesn't matter)
    pub target_direction: model::Vec2,
    /// Order to perform an action, or None
    pub action: Option<model::ActionOrder>,
}

impl trans::Trans for UnitOrder {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.target_velocity.write_to(writer)?;
        self.target_direction.write_to(writer)?;
        self.action.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let target_velocity: model::Vec2 = trans::Trans::read_from(reader)?;
        let target_direction: model::Vec2 = trans::Trans::read_from(reader)?;
        let action: Option<model::ActionOrder> = trans::Trans::read_from(reader)?;
        Ok(Self {
            target_velocity,
            target_direction,
            action,
        })
    }
}